Make route handler filters work for MVC controllers

davidfowl opened the pull request:
- This change makes route handler filters work better with MVC controller actions by wrapping the action delegate directly. This means that run after action filters.
- To do so we need to expose the RouteHandlerFilters on the RouteEndpointBuilder and use them in MVC

Fixes https://github.com/dotnet/aspnetcore/issues/42799
Fixes https://github.com/dotnet/aspnetcore/issues/42593

davidfowl commented on src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs:
```diff
@@ -406,6 +413,35 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string?>
+            // Most inner delegate needs a ControllerRouteHandlerInvocationContext to function
+            RouteHandlerFilterDelegate del = static async (context) =>
+            {
+                var controllerRhiContext = (ControllerRouteHandlerInvocationContext)context;
```
@captainsafia unlike the RequestDelegateFactory, this needs to be of a specific type which makes me think we should fail if you replace the instance since it won't work in all cases.

davidfowl committed:
Make route handler filters work for MVC controllers
    - This change makes route handler filters work better with MVC controller actions by wrapping the action delegate directly.
    - To do so we need to expose the RouteHandlerFilters on the RouteEndpointBuilder and use them in MVC
    - Implement GetEndpointGroup on ActionEndpointDataSource
    - Reuse the action method executor logic to pick the best invoker even in the filter scenario.
    - Added tests

Kahbazi commented on src/Mvc/Mvc.Core/src/Infrastructure/ActionMethodExecutor.cs:
```diff
@@ -48,17 +52,65 @@ public static ActionMethodExecutor GetExecutor(ObjectMethodExecutor executor)
         {
             executor.Execute(controller, arguments);
-            return new ValueTask<IActionResult>(new EmptyResult());
+            return new(new EmptyResult());
```
Is it possible to only use one single instance of `EmptyResult` instead of creating a new instance every time?

davidfowl commented on src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs:
```diff
@@ -406,6 +421,31 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string?>
+            // Make them null to avoid the default behavior
+            reb.FilterFactories.Clear();
+
+            EndpointFilterDelegate del = invocationContext =>
```
```suggestion
            EndpointFilterDelegate del = static invocationContext =>
```

davidfowl commented on src/Mvc/Mvc.RazorPages/test/Infrastructure/PageActionEndpointDataSourceTest.cs:
```diff
@@ -150,15 +150,14 @@ public void GroupedEndpoints_AppliesConventions_RouteSpecificMetadata()
                 Assert.Equal("/group1/test", e.RoutePattern.RawText);
                 Assert.Same(actions[0], e.Metadata.GetMetadata<ActionDescriptor>());
-                // Assert.Equal(new[] { "group", "Hi there", "A" }, e.Metadata.GetOrderedMetadata<string>());
+                Assert.Equal(new[] { "A", "group", "Hi there" }, e.Metadata.GetOrderedMetadata<string>());
```
This is wrong no? The group should be overridden by the local metadata entry?

captainsafia replied:
I had this inquiry as well when I originally wrote this but I eventually deduced that its because I'm registering the action-specific metadata differently in each scenario.

In the controller scenario, the local metadata is registered using a convention on the controller action builder. The route-specific conventions run after the global ones when the action endpoint is created.

In the Razor page scenario, the local metadata is provided in `ActionDescriptor.EndpointMetadata` which gets registered before the metadata that is added before conventions.

https://github.com/dotnet/aspnetcore/blob/c2cba55ca149ab3bd3fadcf97b9d16a74f561111/src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs#L315-L322

Now, to be honest, it seems weird to me that the code above would treat action-specific metadata as having a _lower_  precedence than other conventions, but it seems like this behavior has been around for a while.

For consistencies sake, I can update both tests to use the `Page`/`Controller` convention builders to register metadata.

captainsafia replied:
Actually, based on my understanding of the code, this makes sense for asserting the behavior that group metadata has lower precedence than local metadata. The `AddRoute` method used in the controller tests doesn't necessarily map for Razor pages since it is used to set fallback routes for controllers.

davidfowl replied:
As long as there's a test that verifies that adding metadata via say an attribute on the controller action isn't overridden by what a group adding that same attribute, then we're golden.

halter73 replied:
This makes sense to me. Do we have the test for attributes yet? It might make sense to add a call to `MapGroup` inside `StartupForEndpointFilters` so we can test it end-to-end. This way we can test both adding filters via a group and adding other metadata, and the relative ordering of all of this.

halter73 replied:
Also, I want to make sure any metadata applied to the outer group does have lower precedence than something added via something like `outerGroup.WithMetadatata("Before A").AddRoute("1", "/1/{controller}/{action}/{id?}", null, null, null).Add(b => b.Metadata.Add("A"));`

davidfowl replied:
Added a test here, this took me a while to understand but it seems that route specific entries override attributes on the endpoint.

davidfowl commented on src/Mvc/test/WebSites/RoutingWebSite/StartupForGroups.cs:
```diff
@@ -29,7 +29,11 @@ public virtual void Configure(IApplicationBuilder app)
-            controllerGroup.MapControllers();
+            controllerGroup.MapControllers().AddEndpointFilter((context, next) =>
+            {
+                context.HttpContext.Items[nameof(IEndpointFilter)] = true;
```
Maybe also verify the method info. That's one of the things that changes would be good to assert.

halter73 commented on src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs:
```diff
@@ -246,6 +259,8 @@ public ActionEndpointFactory(RoutePatternTransformer routePatternTransformer, IE
             route.Conventions[i](builder);
         }
 
+        builder.FilterFactories.Clear();
```
Depending on the merge order, either this PR or my PR (#42827) will want to undue this because my PR stops the re-application of filter during Build().

I'm okay merging second, so this is really a note to self.

halter73 replied:
I'm setting #42827 to auto-merge, so it's likely this PR that will have the merge conflict. But it means we can remove this line:

```suggestion
```

halter73 commented on src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs:
```diff
@@ -406,6 +421,28 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string?>
             perRouteConventions[i](builder);
         }
+
+        if (builder is { FilterFactories.Count: > 0 } reb && action is ControllerActionDescriptor cad)
+        {
+            var routeHandlerFilters = reb.FilterFactories;
```
```suggestion
        if (builder.FilterFactories.Count > 0 && action is ControllerActionDescriptor cad)
        {
            var routeHandlerFilters = builder.FilterFactories;
```

halter73 commented on src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs:
```diff
@@ -406,6 +421,28 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string?>
+                del = filterFactory(context, del);
+            }
+
+            cad.FilterDelegate = del;
```
How about we keep a reference to the `originalDelegate` and we did the same thing we do in RequestDelegateFactory.

https://github.com/dotnet/aspnetcore/blob/8ce81e9dac0db17415dd21eea4f02ce025f3924c/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L355-L360

```suggestion
            if (!ReferenceEquals(originalDelegate, del))
            {
                cad.FilterDelegate = del;
            }
```

davidfowl replied:
Why?

halter73 replied:
Why not? It's easy enough, and we already do it elsewhere. I'm not demanding it.

davidfowl replied:
Yea, seems fine as long as it's easy

davidfowl replied:
https://github.com/dotnet/aspnetcore/pull/42557/commits/24a8ea3d92c682255a5006d9b03e50c1a810ddb3

davidfowl committed:
Naming updates

davidfowl committed:
Fixed naming

davidfowl committed:
No need to clean filter factories anymore

davidfowl committed:
Fixed how group conventions run on mvc endpoints
- Make sure conventions are applied in the right order.
- Added tests

davidfowl committed:
Stephen optimization
- If the initial delegate is the same as the final delegate then skip the filter pipeline.

davidfowl commented:
OK this is ready for another pass.

davidfowl committed:
Fixed the grouping tests

captainsafia commented on src/Mvc/Mvc.Core/src/Routing/ActionEndpointFactory.cs:
```diff
@@ -301,17 +324,27 @@ private static (RoutePattern resolvedRoutePattern, IDictionary<string, string?>
         IReadOnlyList<Action<EndpointBuilder>> conventions,
         IReadOnlyList<Action<EndpointBuilder>> perRouteConventions)
     {
+        // REVIEW: The RouteEndpointDataSource adds HttpMethodMetadata before running group conventions
```
I think it's goodness to give access to this metadata to group conventions. I don't anticipate that it'll be as required here as it is in REDS, where it's needed for OpenAPI support. IMO, nice to have but not groundbreaking.

davidfowl replied:
> I think it's goodness to give access to this metadata to group conventions. I don't anticipate that it'll be as required here as it is in REDS, where it's needed for OpenAPI support. IMO, nice to have but not groundbreaking.

Open API fixes it with Finally 😄

halter73 replied:
We should definitely fix this inconsistency. RouteEndpointDataSource definitely has the nicer, more convenient behavior. Maybe we can sneak this into #43225.

captainsafia replied:
@halter73 Sure -- I can adjust. Do any other conventions rely on this particular metadata? OpenAPI was the only one that came to mind.

davidfowl replied:
I think we’d need a scenario where it matters. This code needs to be re-worked to do that I think. We don’t want group conventions to override endpoint metadata in attributes and metadata set from the action descriptor

halter73 replied:
> This code needs to be re-worked to do that I think. We don’t want group conventions to override endpoint metadata in attributes and metadata set from the action descriptor

I agree with this part.

> I think we’d need a scenario where it matters. 

I think consistency is enough of a reason to say we ought to do this. I can agree that fixing this isn't the highest priority thing, and if it's too complicated it might not be worth doing now.

halter73 replied:
And yes, WithOpenApi is the only thing I know of that already depends on this and it *is* fixed by `Finally`.

captainsafia commented on src/Mvc/test/Mvc.FunctionalTests/RoutingGroupsTest.cs:
```diff
@@ -4,7 +4,9 @@
 using System.Net;
 using System.Net.Http;
 using System.Net.Http.Json;
+using System.Text.Json;
```
Still needed?

```suggestion
```
